### DISCUSSION TITLE
Improve inference about which method should be call on type deconstruction

### DISCUSSION THREAD

**Discussion by leandromoh** [dotnet/csharplang#3621] [Language Ideas]
let imagine we have a `Person` type, which has the follow extensions for deconstruction:
```c#
public static void Deconstruct(this Person p, out string name, out int age) =>
    (name, age) = (p.Name, p.Age);

public static void Deconstruct(this Person p, out string name, out string email) =>
    (name, email) = (p.Name, p.Email);
```
obviously, the follow call has ambiguity:

```c#
var p1 = new Person { Name = "bob", Age = 45, Sex = 'M', Email = "email@bob.com" };

var (name, email) = p1;
```
but even specifying the types, it also fail:
```c#
(string name, string email) = p1;
```

I think would be nice to consider a target-typed here :)

**Comment by leandromoh** [comment]
This can be solved easily with the [nominal deconstruction](https://github.com/dotnet/csharplang/issues/3546) proposal

**Comment by HaloFour** [comment]
Here are the relevant meeting notes on how deconstructors are resolved: https://github.com/dotnet/csharplang/blob/master/meetings/2016/LDM-2016-09-06.md#how-do-we-select-deconstruct-methods

**Comment by leandromoh** [comment]
I read the document, but could not understand. I should call deconstruction method myself?
